subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1634 @@
+{
+	int rc;
+	struct zms_ate last_ate, first_ate, close_ate, empty_ate;
+	size_t ate_size;
+	/* Initialize addr to 0 for the case fs->sector_count == 0. This
+	 * should never happen as this is verified in zms_mount() but both
+	 * Coverity and GCC believe the contrary.
+	 */
+	uint64_t addr = 0U;
+	uint32_t i, closed_sectors = 0;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	/* step through the sectors to find a open sector following
+	 * a closed sector, this is where zms can write.
+	 */
+
+	for (i = 0; i < fs->sector_count; i++) {
+		addr = ((uint64_t)i << ADDR_SECT_SHIFT) + (uint32_t)(fs->sector_size - ate_size);
+
+		/* read the first and second ate to find out if sector is
+		 * open/close
+		 */
+		rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+		if (rc) {
+			goto end;
+		}
+
+		/* update cycle count */
+		fs->sector_cycle = empty_ate.cycle_cnt;
+
+		if (zms_close_ate_valid(fs, &close_ate) && zms_empty_ate_valid(fs, &empty_ate) &&
+		    (close_ate.cycle_cnt == empty_ate.cycle_cnt)) {
+			/* closed sector */
+			closed_sectors++;
+			/* Let's check first that we support this ZMS version */
+			if (ZMS_GET_VERSION(empty_ate.metadata) != ZMS_DEFAULT_VERSION) {
+				LOG_ERR("ZMS Version is not supported");
+				rc = -ENOEXEC;
+				goto end;
+			}
+			zms_sector_advance(fs, &addr);
+			/* addr is pointing to the close ATE */
+			rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+			if (rc) {
+				goto end;
+			}
+			/* update cycle count */
+			fs->sector_cycle = empty_ate.cycle_cnt;
+			if (!zms_empty_ate_valid(fs, &empty_ate) ||
+			    !zms_close_ate_valid(fs, &close_ate) ||
+			    (close_ate.cycle_cnt != empty_ate.cycle_cnt)) {
+				/* open sector */
+				break;
+			}
+		}
+	}
+	/* all sectors are closed, this is not a zms fs */
+	if (closed_sectors == fs->sector_count) {
+		rc = -EDEADLK;
+		goto end;
+	}
+
+	if (i == fs->sector_count) {
+		/* none of the sectors were closed, which means that the first
+		 * sector is the one in use, except if there are only 2 sectors.
+		 * Let's check if the last sector has valid ATEs otherwise set
+		 * the open sector to the first one.
+		 */
+		rc = zms_flash_ate_rd(fs, addr - 2 * ate_size, &first_ate);
+		if (rc) {
+			goto end;
+		}
+		if (!zms_ate_valid(fs, &first_ate)) {
+			zms_sector_advance(fs, &addr);
+		}
+		rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+		if (rc) {
+			goto end;
+		}
+		if (!zms_empty_ate_valid(fs, &empty_ate)) {
+			if (ebw_required) {
+				zms_flash_erase_sector(fs, addr);
+			}
+			zms_add_empty_ate(fs, addr);
+		} else {
+			/* empty_ate is valid let's check the version */
+			if (ZMS_GET_VERSION(empty_ate.metadata) != ZMS_DEFAULT_VERSION) {
+				LOG_ERR("ZMS Version is not supported");
+				rc = -ENOEXEC;
+				goto end;
+			}
+		}
```

Comment:
Reverse the logic: expect the valid system first.
Then move the comment 1024-1026 to the `!zms_empty_ate_valid` block first, or better make "zms_format()" function and make it do the magic, so that it will be clear that we are making new system here.

Additionally, would it be possible to parametrize the zms_init function to take a parameter that would allow to:
 - just return on error when there is no FS found?
 - force re-initialization, just because we want to get rid of what is on the partition and set a new fs anyway (we know what we are doing)
 
That should be possibly passed from `zms_mount` or through context.

Reply:
Actually there is nothing to do here if the empty_ate is valid, magic is valid and version is correct
The comment that you mentioned is about a special case where there are either 2 open sectors or more than 2 but all of them are open.
At the end of the first loop we are pointing to the last sector of the storage.
The decision that we need to make here is the following :
When there is no valid ATE in the last sector it means that either we are in the scenario of 2 sectors but the currently used one is sector 0, or we have more than 2 sectors which means also that the used sector is sector 0
After advancing to the sector 0 or remaining in sector 1, we must be sure that VERSION and MAGIC exist.
If MAGIC doesn't exist => storage never used (we must continue)
If MAGIC exist but VERSION is wrong then we return on error.
I think that it is a good idea to add a parameter to the init to force formatting the storage if something went wrong.
I will add it later (not in this PR)